File: src/telegram.rs
use worker::*;
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize)]
struct TelegramMessage {
    chat_id: String,
    text: String,
    parse_mode: String,
}

#[derive(Debug, Deserialize)]
struct TelegramResponse {
    ok: bool,
    description: Option<String>,
}

pub async fn send_telegram_notification(
    bot_token: &str,
    chat_id: &str,
    message: &str,
) -> Result<()> {
    let url = format!("https://api.telegram.org/bot{}/sendMessage", bot_token);
    
    let telegram_message = TelegramMessage {
        chat_id: chat_id.to_string(),
        text: format!("🤖 Claude Code Notification\n\n{}", message),
        parse_mode: "HTML".to_string(),
    };
    
    let mut init = RequestInit::new();
    init.with_method(Method::Post);
    
    let mut headers = Headers::new();
    headers.set("Content-Type", "application/json")?;
    init.with_headers(headers);
    
    let body = serde_json::to_string(&telegram_message).unwrap();
    init.with_body(Some(body.into()));
    
    let request = Request::new_with_init(&url, &init)?;
    
    let mut response = Fetch::Request(request).send().await?;
    let response_text = response.text().await?;
    
    let telegram_response: TelegramResponse = serde_json::from_str(&response_text)
        .map_err(|e| Error::RustError(format!("Failed to parse Telegram response: {}", e)))?;
    
    if !telegram_response.ok {
        return Err(Error::RustError(format!(
            "Telegram API error: {}",
            telegram_response.description.unwrap_or_else(|| "Unknown error".to_string())
        )));
    }
    
    Ok(())
}

File: src/telegram.ts
import type { ApiResponse, Message } from "@grammyjs/types";
import { z } from "zod";

type TelegramMessageReceipt = {
	chat: Pick<Message["chat"], "id">;
};

type TelegramSendMessageApiResponse = ApiResponse<TelegramMessageReceipt>;

const telegramSendMessageApiResponseSchema = z.discriminatedUnion("ok", [
	z.object({
		ok: z.literal(true),
		result: z.object({
			chat: z.object({ id: z.number() }),
		}),
	}),
	z.object({
		ok: z.literal(false),
		error_code: z.number(),
		description: z.string(),
	}),
]) satisfies z.ZodType<TelegramSendMessageApiResponse>;

export type TelegramMessageDeliveryOutcome =
	| { status: "sent"; chatId: Message["chat"]["id"] }
	| { status: "rejected"; errorCode: number; description: string };

export interface TelegramMessageRequest {
	botToken: string;
	chatId: number | string;
	text: string;
	parseMode?: "Markdown" | "HTML";
	disableNotification?: boolean;
}

export type TelegramHttpClient = (
	input: RequestInfo | URL,
	requestInitialization?: RequestInit,
) => Promise<Response>;

export function parseTelegramSendMessageApiResponse(
	rawApiResponse: unknown,
): TelegramSendMessageApiResponse {
	const parseResult =
		telegramSendMessageApiResponseSchema.safeParse(rawApiResponse);
	if (!parseResult.success) {
		throw new Error(
			"Telegram API returned a response that does not match the expected contract",
		);
	}

	return parseResult.data;
}

export async function sendTelegramMessage(
	messageRequest: TelegramMessageRequest,
	telegramHttpClient: TelegramHttpClient = fetch,
): Promise<TelegramMessageDeliveryOutcome> {
	const telegramApiUrl = `https://api.telegram.org/bot${messageRequest.botToken}/sendMessage`;
	const telegramApiResponse = await telegramHttpClient(telegramApiUrl, {
		method: "POST",
		headers: {
			"Content-Type": "application/json",
		},
		body: JSON.stringify({
			chat_id: messageRequest.chatId,
			text: messageRequest.text,
			parse_mode: messageRequest.parseMode,
			disable_notification: messageRequest.disableNotification,
		}),
	});
	const rawApiResponse: unknown = await telegramApiResponse.json();
	const parsedApiResponse = parseTelegramSendMessageApiResponse(rawApiResponse);

	if (!parsedApiResponse.ok) {
		return {
			status: "rejected",
			errorCode: parsedApiResponse.error_code,
			description: parsedApiResponse.description,
		};
	}

	return {
		status: "sent",
		chatId: parsedApiResponse.result.chat.id,
	};
}
